packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx

```diff
@@ -16,54 +30,129 @@ type RemoteBlockProps = {
     sourceUrl,
     crossFrame,
     onBlockLoaded,
   );
 
+  const wrapperRef = React.useRef<HTMLDivElement>(null);
+
+  const { graphService } = useGraphEmbedderService(wrapperRef, {
+    callbacks: graphCallbacks,
+    ...graphProperties,
+  });
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.blockEntity({ data: graphProperties.blockEntity });
+    }
+  }, [graphProperties.blockEntity, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.blockGraph({ data: graphProperties.blockGraph });
+    }
+  }, [graphProperties.blockGraph, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.entityTypes({ data: graphProperties.entityTypes });
+    }
+  }, [graphProperties.entityTypes, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.linkedAggregations({
+        data: graphProperties.linkedAggregations,
+      });
+    }
+  }, [graphProperties.linkedAggregations, graphService]);
+
   if (loading) {
     return <BlockLoadingIndicator />;
   }
 
-  if (!Component) {
+  if (!blockSource) {
     throw new Error("Could not load and parse block from URL");
   }
 
   if (err) {
     throw err;
   }
 
-  if (typeof Component === "string") {
-    /**
-     * This HTML block has no props available to it, unless loaded via FramedBlock.
-     * @todo do something about this. throw if not in an iframe?
-     *    or check for iframe status and assign props to window here, not FramedBlock?
-     */
-    return (
-      <HtmlBlock
-        html={Component}
-        blockFunctions={blockFunctions}
-        blockProperties={blockProperties}
-      />
+  const propsToInject: BlockGraphProperties<Record<string, any>> & {
+    editableRef: any;
+  } = {
+    editableRef,
+    graph: graphProperties,
+  };
+
+  // @todo remove this React default when we update all blocks to 0.2
+  const entryPoint = blockMetadata.blockType?.entryPoint ?? "react";
+
+  if (entryPoint === "html" && typeof blockSource !== "string") {
+    throw new Error(
+      `'html' entryPoint expects source to be typeof 'string', but got: ${typeof blockSource}`,
+    );
+  } else if (entryPoint === "custom-element") {
+    if (!(blockSource instanceof HTMLElement)) {
+      throw new Error(
+        `'custom-element' entryPoint expects parsed source to be instanceof 'HTMLElement'`,
+      );
+    }
+    if (typeof blockMetadata.blockType.tagName !== "string") {
+      throw new Error(
+        `Must provide blockType.tagName when entryPoint is 'custom-element'`,
+      );
+    }
+  } else if (entryPoint === "react" && typeof blockSource !== "function") {
+    throw new Error(
+      `'react' entryPoint expects parsed source to be typeof 'function', but got: ${typeof blockSource}`,
     );
+  } else if (!["html", "custom-element", "react"].includes(entryPoint)) {
+    throw new Error(`Invalid entryPoint '${entryPoint}'`);
   }
```

Comment:
Nit: I could be missing something, but my instinct this logic would probably be better encapsulated within the `BlockRenderer` component? It would probably lead to less casting (as you can directly `return` the JSX elements in each if conditional), and removes the need for the `customElement`/`html`/`ReactComponent` props.

Definitely doesn't block merging, just a thought I'd mention it as it could potentially simplify the logic/prop passing here.

Reply:
Thanks @benwerner01 , good suggestion - I've addressed in https://github.com/hashintel/hash/pull/752/commits/5e68957e24b71d5e58c3725778e8321d06682362

There's still a couple of casts in here which could be fixed but I think they're okay for now

Reply:
Agreed 👍